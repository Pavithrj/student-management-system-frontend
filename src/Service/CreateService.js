export async function getStudents() {
    const response = await fetch("http://localhost:4000/students");
    const data = await response.json();
    return data;
}

export async function createStudents(student) {
    const response = await fetch("http://localhost:4000/students", {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(student),
    });
    if (response.status === 400) {
        const error = await response.json();
        throw error;
    } else {
        const data = await response.json();
        return data;
    }
}