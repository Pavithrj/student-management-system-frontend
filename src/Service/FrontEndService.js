import { API_BASE_URL } from "../constants";

export async function getStudents() {
    const response = await fetch(`${API_BASE_URL}/students`);
    const data = await response.json();
    return data;
}

export async function createStudent(student) {
    const response = await fetch(`${API_BASE_URL}/create`, {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(student),
    });

    if (response.status === 400) {
        const error = await response.json();
        throw error;
    } else {
        const data = await response.json();
        return data;
    }
}

export async function getStudentDetails(rollNo) {
    const response = await fetch(`${API_BASE_URL}/create/students/${rollNo}`, {
        method: 'GET',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
    });

    if (response.status === 400) {
        const error = await response.json();
        throw error;
    } else {
        const data = await response.json();
        return data;
    }
}

export async function deleteStudent(rollNo) {
    const response = await fetch(`${API_BASE_URL}/students/${rollNo}`, {
        method: 'DELETE',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
    });

    if (response.status === 400) {
        const error = await response.json();
        throw error;
    } else {
        const data = await response.json();
        return data;
    }
}

export async function editStudent(rollNo, student) {
    try {
        const response = await fetch(`${API_BASE_URL}/students/edit/${rollNo}`, {
            method: 'PUT',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(student)
        });

        if (!response.ok) {
            throw new Error("Error");
        }
    }
    catch (error) {
        console.log("error", error);
    }
}