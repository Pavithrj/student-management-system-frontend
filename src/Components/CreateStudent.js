import React, { useState } from "react";
import { useNavigate } from 'react-router-dom';
import { createStudent } from "../Service/FrontEndService";
import './CreateStudent.css';
import DoneIcon from '@mui/icons-material/Done';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';

function CreateStudent() {
    const [createName, setCreateName] = useState("");
    const [createAge, setCreateAge] = useState("");
    const [createRollNo, setCreateRollNo] = useState("");
    const [errorMessage, setErrorMessage] = useState("");
    const navigate = useNavigate();

    const handleInputChange = (e, stateSetter) => {
        stateSetter(e.target.value);
    };

    const GoBack = () => {
        navigate('/');
    };

    const SubmitCreatedStudent = async () => {
        const newStudent = {
            name: createName,
            age: createAge,
            rollNo: createRollNo
        };

        try {
            await createStudent(newStudent);
            navigate('/', { state: { isSuccessful: true } });
        } catch (error) {
            setErrorMessage(error.message);
            console.log("Error in calling the function.", error);
        }
    };

    const renderInputField = (label, value, onChange) => {
        return (
            <div>
                <label>
                    {label}:
                    <input value={value} onChange={onChange} aria-label={label} />
                </label>
            </div>
        );
    };

    return (
        <div>
            <p className="error-message">{errorMessage}</p>
            <h1>Create Student:</h1>
            <div className="create-student-details-input">
                {renderInputField("Name", createName, (e) => handleInputChange(e, setCreateName))}
                {renderInputField("Age", createAge, (e) => handleInputChange(e, setCreateAge))}
                {renderInputField("RollNo", createRollNo, (e) => handleInputChange(e, setCreateRollNo))}
            </div>
            <br />
            <div className="create-student-buttons">
                <button data-testid="create-student-go-back-btn" onClick={GoBack}><ArrowBackIcon /></button>
                <button data-testid="doneButton" onClick={SubmitCreatedStudent}><DoneIcon /></button>
            </div>
        </div>
    )
}

export default CreateStudent;