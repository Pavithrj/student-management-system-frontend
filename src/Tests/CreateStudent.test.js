import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useNavigate } from 'react-router-dom';
import { createStudent } from '../Service/FrontEndService';
import CreateStudent from '../Components/CreateStudent';

jest.mock('react-router-dom', () => ({
    useNavigate: jest.fn(),
}));

jest.mock('../Service/FrontEndService', () => ({
    createStudent: jest.fn(),
}));

describe('CreateStudent Component', () => {
    beforeEach(() => {
        useNavigate.mockClear();
        createStudent.mockClear();
    });

    it('should render the component', () => {
        const { asFragment } = render(<CreateStudent />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should navigate back to home page when GoBack button is clicked', () => {
        const navigate = jest.fn();
        useNavigate.mockReturnValue(navigate);

        render(<CreateStudent />);

        const goBackButton = screen.getByTestId('goBackButton');
        fireEvent.click(goBackButton);

        expect(navigate).toHaveBeenCalledWith('/');
    });

    it('should submit the created student when Done button is clicked', async () => {
        const navigate = jest.fn();
        useNavigate.mockImplementation(() => navigate);

        render(<CreateStudent />);

        const nameInput = screen.getByLabelText('Name:');
        const ageInput = screen.getByLabelText('Age:');
        const rollNoInput = screen.getByLabelText('RollNo:');

        fireEvent.change(nameInput, { target: { value: 'John Doe' } });
        fireEvent.change(ageInput, { target: { value: '25' } });
        fireEvent.change(rollNoInput, { target: { value: '123' } });

        const doneButton = screen.getByTestId('doneButton');
        fireEvent.click(doneButton);

        expect(createStudent).toHaveBeenCalledWith({
            name: 'John Doe',
            age: '25',
            rollNo: '123',
        });

        await screen.findByText('Create Student:');

        expect(navigate).toHaveBeenCalledWith('/', { state: { isSuccessful: true } });
    });

    it('should display error message when createStudent function throws an error', async () => {
        const errorMessage = 'Failed to create student';
        createStudent.mockRejectedValueOnce(new Error(errorMessage));

        render(<CreateStudent />);

        const doneButton = screen.getByTestId('doneButton');
        fireEvent.click(doneButton);

        await screen.findByText(errorMessage, {}, { timeout: 3000 });

        expect(screen.getByText(errorMessage)).toBeInTheDocument();
    });
});