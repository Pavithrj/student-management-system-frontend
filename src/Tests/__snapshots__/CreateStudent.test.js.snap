// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateStudent Component should render the component 1`] = `
<DocumentFragment>
  <div>
    <p
      class="error-message"
    />
    <h1>
      Create Student:
    </h1>
    <div
      class="create-student-details-input"
    >
      <div>
        <label>
          Name:
          <input
            aria-label="Name"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          Age:
          <input
            aria-label="Age"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          RollNo:
          <input
            aria-label="RollNo"
            value=""
          />
        </label>
      </div>
    </div>
    <br />
    <div
      class="create-student-buttons"
    >
      <button
        data-testid="create-student-go-back-btn"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </button>
      <button
        data-testid="doneButton"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DoneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
